### PR TITLE
api: add diagnose api

### PR DESCRIPTION
**Pull request by jomenxiao**
diagnose PD and give some advice for the maintainer

### CODE REVIEW

**Comment by siddontang** [comment]
please use a better title and add a description for your PR.

**Comment by disksing** [comment]
PTAL @nolouch @Connor1996

File: server/api/diagnose.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		d = diagnoseMap[tikvCap70]
+	case tikvCap80:
+		d = diagnoseMap[tikvCap80]
+	case tikvCap90:
+		d = diagnoseMap[tikvCap90]
+	case tikvLostPeers:
+		d = diagnoseMap[tikvLostPeers]
+	case tikvLostPeersLongTime:
+		d = diagnoseMap[tikvLostPeersLongTime]
+	}
```
Comment (disksing) [inline server/api/diagnose.go]: I don't think the switch-case is necessary.

File: server/api/diagnose.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+type Recommended struct {
+	Module      string `json:"module"`
+	Description string `json:"description"`
+	Instruction string `json:"instruction"`
+	Level       string `json:"level"`
+}
+
+func diagnosePD(key diagnoseType) *Recommended {
+	mapLock.RLock()
+	defer mapLock.RUnlock()
```
Comment (disksing) [inline server/api/diagnose.go]: The readonly map doesn't need mutex.

File: server/api/diagnose.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/pd/server"
+	"github.com/unrolled/render"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type diagnoseType int
+type diagnoseKV map[diagnoseType][4]string
```
Comment (disksing) [inline server/api/diagnose.go]: Seems we can make it _map[type]Recommanded_.

File: server/api/diagnose.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		rd:  rd,
+	}
+}
+
+//Recommended return detail
+type Recommended struct {
+	Module      string `json:"module"`
+	Description string `json:"description"`
+	Instruction string `json:"instruction"`
+	Level       string `json:"level"`
```
Comment (disksing) [inline server/api/diagnose.go]: Put _Level_ before _Description_ and _Instruction_.

File: server/api/diagnose.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	return &Recommended{
+		Module:      d[0],
+		Level:       d[1],
+		Description: d[2],
+		Instruction: d[3],
+	}
+}
+
+func (d *diagnoseHandler) membersDiagnose(rd []*Recommended) error {
```
Comment (disksing) [inline server/api/diagnose.go]: Should return `rd`, append it in the function won't work.

File: server/api/utiletcdhttpapi.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+func getEtcdPeerStats(etcdClientURL string) (*PeerStats, error) {
+	ps := &PeerStats{}
+	resp, err := doGet(fmt.Sprintf("%s%s", etcdClientURL, etcdPeerStatsAPI))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := readJSON(resp.Body, ps); err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (nolouch) [inline server/api/utiletcdhttpapi.go]: close resp.Body

Comment (jomenxiao) [reply server/api/utiletcdhttpapi.go]: resp.Body is close in `readJSON`

File: server/api/diagnose.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/pingcap/pd/server"
+	"github.com/unrolled/render"
+	"net/http"
+	"time"
+)
+
+type diagnoseType int
+type diagnoseKV map[diagnoseType]Recommended
+
+//Recommended return detail
```
Comment (disksing) [inline server/api/diagnose.go]: `// Recommended is ...`
btw, `Recommended` is not a good name. Better use noun form `recommendation`, or maybe `Analytic`.

File: server/api/diagnose.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+type Recommended struct {
+	Module      string `json:"module"`
+	Level       string `json:"level"`
+	Description string `json:"description"`
+	Instruction string `json:"instruction"`
+}
+
+const (
+	//analyze level
```
Comment (disksing) [inline server/api/diagnose.go]: Add a space between `//` and the comment after it.

Comment (disksing) [reply server/api/diagnose.go]: level -> levels.

File: server/api/diagnose.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	lenMembers = len(members.Members)
+	if lenMembers > 0 {
+		for _, m := range members.Members {
+			pm, err := getEtcdPeerStats(m.ClientUrls[0])
+			if err != nil {
+				//get peer etcd failed
+				lostMembers++
+				continue
+			}
+			if time.Now().Sub(pm.LeaderInfo.StartTime) < time.Duration(3*time.Minute) {
```
Comment (Connor1996) [inline server/api/diagnose.go]: use time.Since() instead, which is shorthand for time.Now().Sub()

File: server/api/diagnose.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	levelMajor    = "Major"
+	levelCritical = "Critical"
+	levelFatal    = "Fatal"
+
+	//analyze module
+	modMember   = "member"
+	modTiKV     = "TiKV"
+	modReplica  = "Replic"
+	modSchedule = "Schedule"
+	modDefault  = "indefinite"
```
Comment (disksing) [inline server/api/diagnose.go]: `indefinite` looks strange, why not just `Default`?

File: server/api/diagnose.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	memberLeaderChanged
+	tikvCap70
+	tikvCap80
+	tikvCap90
+	tikvLostPeers
+	tikvLostPeersLongTime
+)
+
+var (
+	diagnoseMap = diagnoseKV{
```
Comment (disksing) [inline server/api/diagnose.go]: We can just `diagnoseMap = map[diagnoseType]Recommended {...}`. No need to introduce a new type.

File: server/api/diagnose.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func newDiagnoseHandler(svr *server.Server, rd *render.Render) *diagnoseHandler {
+	return &diagnoseHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+func diagnosePD(key diagnoseType) *Recommended {
```
Comment (disksing) [inline server/api/diagnose.go]: I think we can be sure the key is valid, so the function is unnecessary. It is ok to use `disgnoseMap[key]` directly.

File: server/api/utiletcdhttpapi.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/juju/errors"
+)
+
+const (
+	//etcd peer detail API
+	//return struct https://github.com/coreos/etcd/blob/master/etcdserver/stats/server.go
+	etcdPeerStatsAPI = "/v2/stats/self"
+)
+
+//PeerStats show peer stats
```
Comment (disksing) [inline server/api/utiletcdhttpapi.go]: `// PeerStats is ...`.

File: server/api/diagnose.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	return nil
+}
+
+func (d *diagnoseHandler) tikvDiagnose(rdd *[]*Recommended) error {
+	return nil
+}
+
+func (d *diagnoseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rdd := []*Recommended{}
```
Comment (Connor1996) [inline server/api/diagnose.go]: I prefer `membersDiagnose()` return `rdd`, there is no need to define `rdd` here.

File: server/api/diagnose_test.go [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (s *testDiagnoseAPISuite) SetUpSuite(c *C) {
+	s.hc = newHTTPClient()
+}
+
+func checkDiagnoseResponse(c *C, body []byte) {
+	got := []Recommended{}
+	json.Unmarshal(body, &got)
+	for _, r := range got {
```
Comment (disksing) [inline server/api/diagnose_test.go]: It seems the response will be empty for a cluster with no abnormal condition, so the test will not check anything.

File: server/api/diagnose.go (hunk 14 above)
Comment (jomenxiao) [reply server/api/diagnose.go]: only `membersDiagnose` now,
will add other functions (like `tikvDiagnose`)

File: server/api/diagnose.go (hunk 12 above)
Comment (jomenxiao) [reply server/api/diagnose.go]: this function will add more params,like ` diagnosePD(key diagnoseType, descAdd, instAdd string)`
`Description` and `Description` support additional messages.

File: server/api/diagnose_test.go (hunk 15 above)
Comment (jomenxiao) [reply server/api/diagnose_test.go]: `follow.Close()` ?

File: server/api/diagnose.go [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+func newDiagnoseHandler(svr *server.Server, rd *render.Render) *diagnoseHandler {
+	return &diagnoseHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+func diagnosePD(key diagnoseType, descAdd, instAdd string) *Recommendation {
+	d, ok := diagnoseMap[key]
+	if !ok {
```
[the remark this answers is not in the document]
Comment (jomenxiao) [reply server/api/diagnose.go]: now it's always true.
it returns default information for some undefined content.

File: server/api/diagnose_test.go [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+		c.Assert(len(r.Module) != 0, IsTrue)
+		c.Assert(len(r.Level) != 0, IsTrue)
+		c.Assert(len(r.Description) != 0, IsTrue)
+		c.Assert(len(r.Instruction) != 0, IsTrue)
+	}
+}
+
+func (s *testDiagnoseAPISuite) TestDiagnoseSlice(c *C) {
+	_, svrs, clean := mustNewCluster(c, 3)
```
[the remark this answers is not in the document]
Comment (jomenxiao) [reply server/api/diagnose_test.go]: I think to create 3 nodes and `follow.Close()` 
It's a good choice to check the response.

**Pull request merged**